runtime-components/non-repudiation-client/src/main/java/com/daml/nonrepudiation/client/SigningInterceptor.java

```diff
@@ -0,0 +1,80 @@
+
+import com.google.common.io.BaseEncoding;
+import io.grpc.*;
+
+import java.security.KeyPair;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+
+public final class SigningInterceptor implements ClientInterceptor {
+
+    private final PrivateKey key;
+    private final String fingerprint;
+    private final String algorithm;
+
+    private static final Metadata.Key<String> signatureKey =
+            Metadata.Key.of("signature", Metadata.ASCII_STRING_MARSHALLER);
+
+    private static final Metadata.Key<String> algorithmKey =
+            Metadata.Key.of("algorithm", Metadata.ASCII_STRING_MARSHALLER);
+
+    private static final Metadata.Key<String> fingerprintKey =
+            Metadata.Key.of("fingerprint", Metadata.ASCII_STRING_MARSHALLER);
+
+    public SigningInterceptor(KeyPair keyPair, String signingAlgorithm) throws NoSuchAlgorithmException {
+        super();
+        this.key = keyPair.getPrivate();
+        this.algorithm = signingAlgorithm;
+        MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
+        byte[] publicKeyFingerPrint = sha256.digest(keyPair.getPublic().getEncoded());
+        this.fingerprint = BaseEncoding.base64().encode(publicKeyFingerPrint);
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
+                                                               CallOptions callOptions, Channel next) {
+        ClientCall<ReqT, RespT> call = next.newCall(method, callOptions);
+        return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(call) {
+            private Listener<RespT> responseListener = null;
+            private Metadata headers = null;
+            private boolean started = false;
+            private int requested = 0;
+
+            @Override
+            public void start(Listener<RespT> responseListener, Metadata headers) {
+                // Delay start until we have the message body since
+                // the signature in the Metadata depends on the body.
+                this.responseListener = responseListener;
+                this.headers = headers;
+            }
+
+            @Override
+            public void request(int numMessages) {
+                // Delay until we have the message body since the
+                // signature in the Metadata depends on the body.
+                requested += numMessages;
+            }
+
+            @Override
+            public void sendMessage(ReqT request) {
+                byte[] requestBytes = ByteMarshaller.INSTANCE.parse(method.getRequestMarshaller().stream(request));
+                String signature = Base64Signature.sign(algorithm, key, requestBytes);
+                headers.put(signatureKey, signature);
+                headers.put(algorithmKey, algorithm);
+                headers.put(fingerprintKey, fingerprint);
+                if (!started) {
```

Comment:
I think I added this but I don’t understand it anymore. Will `sendMessage` every be called multiple times?

Reply:
Honestly I did not put too much attention to the client, I'll have a second look.

Reply:
If my understanding is correct, I believe a new listener is created for each call, this this guard is not necessary.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/8608/commits/99819c1f8437de5877cbbfb63eccc98fdf9b6588